Reference/Searching/Examine/transforming-index-values.md

```diff
@@ -0,0 +1,120 @@
+---
+versionFrom: 8.0.0
+---
+
+# TransformingIndexValues
+
+In the [Quick Start](Quick-Start/index.md) documentation you can see how to perform a simple search with Examine. That is great if you want to search between node names or you know that you always want to search for a specific field - e.g. `bodyText`. 
+
+However, what if you want to search through several different node types and search across many different fields, you likely will not want to have a query that looks like this:
+
+```cs
+var results = searcher.CreateQuery("content").Field("nodeName", searchTerm)
+                    Or().Field("bodyText", searchTerm)
+                    Or().Field("description", searchTerm)
+                    Or().Field("about", searchTerm)
+                    Or().Field("otherText", searchTerm)
+                    .Execute();
+```
+
+This quickly becomes unmanageable, would have to add each new field to the query as you make them. 
+
+Instead you can use **TransformingIndexValues** (used to be called GatheringNodeData in Umbraco 7). TransformingIndexValues is an Examine event where a new field is created in the index that contains the data from several other fields, you can then search on just that field. This is done via a [composer](../../../Implementation/Composing/index.md).
+
+## Creating a TransformingIndexValues event
+
+:::note
+This example will build upon the Umbraco Starterkit as it is a good starting point for some content. Feel free to use your own site, but some content may need to be generated.
+:::
+
+So to add a TransformingIndexValues event we will add a controller that inherits from `IComponent`. Something like this:
+
+```cs
+public class ExamineEvents : IComponent
+{
+    private readonly IExamineManager _examineManager;
+    private readonly ILogger _logger;
+
+    public ExamineEvents(IExamineManager examineManager, ILogger logger)
+    {
+        _examineManager = examineManager;
+        _logger = logger;
+    }
+
+    public void Initialize()
+    {
+        if (!_examineManager.TryGetIndex("ExternalIndex", out IIndex index))
+            throw new InvalidOperationException("No index found by name ExternalIndex");
+
+        //we need to cast because BaseIndexProvider contains the TransformingIndexValues event
+        if (!(index is BaseIndexProvider indexProvider))
+            throw new InvalidOperationException("Could not cast");
+
+        indexProvider.TransformingIndexValues += IndexProviderTransformingIndexValues;
+    }
+
+    private void IndexProviderTransformingIndexValues(object sender, IndexingItemEventArgs e)
+    {
+        //todo
+    }
+
+    public void Terminate()
+    {
+    }
+}
+```
+
+If this syntax is confusing you can read more about [Components](../../../Implementation/Composing/index.md). We can now add the logic to combine fields in the `IndexProviderTransformingIndexValues` method:
+
+```cs
+if (e.ValueSet.Category == IndexTypes.Content)
+{
+    try
```

Comment:
This is a huge try/catch block... but why? Large try/catch blocks are bad practice and lead to developers thinking that this is a fail safe to proper error checking, this also leads to errors being swallowed and problems will go on silently in the background with developers not knowing what is happening. It's better to let unhandled errors be thrown so people actually know what the problems are.

To do proper error checking you need to determine what can go wrong in this block of code:

* Creating an object `var combinedFields = new StringBuilder();` ... cannot throw, so no try/catch needed
* Iterating over this object `foreach (var fieldValues in e.ValueSet.Values)` ... potentially there could be a null check in there, so need to figure out if `e` or `e.ValueSet` or `e.ValueSet.Values` could ever be null ... I don't think they can so no try/catch needed
* Iterating over this object `foreach (var value in fieldValues.Value)`, just like the above, only a null reference exception could occur if `fieldValues.Value` could ever be null... I don't think it can so no need for try/catch
* `combinedFields.AppendLine(value.ToString());` ... you are already checking for null above, so no exception can occur, no need for a try/catch
* `e.ValueSet.Add("combinedField", combinedFields.ToString());` , we already decided that `e.ValueSet` cannot be null and we definitely know that `combinedFields` cannot be null because we created it. However, an exception can be thrown here if the "combinedField" key already exists in this dictionary, this could happen if someone created a property type called "combinedField", so this should be properly error checked which is simple by just changing the code to `e.ValueSet.TryAdd("combinedField", combinedFields.ToString());` which will do it's own error checking.

We should then remove this try catch so we aren't promote the use of large try/catch/swallow code blocks

Reply:
Great point, fixed it now 👍